sw/device/lib/base/mmio.c

```diff
@@ -4,6 +4,116 @@
 
 #include "sw/device/lib/base/mmio.h"
 
+#include "sw/device/lib/base/memory.h"
+
+/**
+ * Computes how many bytes `addr` is ahead of the previous 32-bit word alignment
+ * boundary.
+ */
+static ptrdiff_t misalignment32_of(uintptr_t addr) {
+  return (addr & (alignof(uint32_t) - 1));
+}
+
+/**
+ * Copies a block of memory between MMIO and main memory while ensuring that
+ * MMIO accesses are word-aligned.
+ *
+ * If `from_mmio` is true, data is copied from MMIO to main memory. Otherwise,
```

Comment:
Nit: Maybe note for future readers why this isn't two functions? (Namely, that those functions would have a lot of finicky dupped code.)

Reply:
Sure, done.